pkg/monitor/monitor.go

```diff
@@ -1,227 +1,148 @@
+		if err := additionalEventIntervalRecorder(ctx, m, m.adminKubeConfig, backend.ExternalLoadBalancerType); err != nil {
+			return err
 		}
-		recordedResource[key] = toStore
-		return
 	}
 
-	existingMetadata, _ := meta.Accessor(existingResource)
-	// without metadata, just stomp in the new value, we can't add annotations
-	if existingMetadata == nil {
-		recordedResource[key] = toStore
-		return
+	// read the state of the cluster apiserver client access issues *before* any test (like upgrade) begins
+	intervals, err := apiserveravailability.APIServerAvailabilityIntervalsFromCluster(client, time.Time{}, time.Time{})
+	if err != nil {
+		klog.Errorf("error reading initial apiserver availability: %v", err)
 	}
+	m.AddIntervals(intervals...)
 
-	existingAnnotations := existingMetadata.GetAnnotations()
-	if existingAnnotations == nil {
-		existingAnnotations = map[string]string{}
-	}
-	existingUpdateCountStr := existingAnnotations[monitorapi.ObservedUpdateCountAnnotation]
-	if existingUpdateCount, err := strconv.ParseInt(existingUpdateCountStr, 10, 32); err != nil {
-		newAnnotations[monitorapi.ObservedUpdateCountAnnotation] = "1"
-	} else {
-		newAnnotations[monitorapi.ObservedUpdateCountAnnotation] = fmt.Sprintf("%d", existingUpdateCount+1)
-	}
+	startPodMonitoring(ctx, m, client)
+	startNodeMonitoring(ctx, m, client)
+	startEventMonitoring(ctx, m, client)
+	shutdown.StartMonitoringGracefulShutdownEvents(ctx, m, client)
 
-	// set the recreate count. increment if the UIDs don't match
-	existingRecreateCountStr := existingAnnotations[monitorapi.ObservedUpdateCountAnnotation]
-	if existingMetadata.GetUID() != newMetadata.GetUID() {
-		if existingRecreateCount, err := strconv.ParseInt(existingRecreateCountStr, 10, 32); err != nil {
-			newAnnotations[monitorapi.ObservedRecreationCountAnnotation] = existingRecreateCountStr
-		} else {
-			newAnnotations[monitorapi.ObservedRecreationCountAnnotation] = fmt.Sprintf("%d", existingRecreateCount+1)
-		}
-	} else {
-		newAnnotations[monitorapi.ObservedRecreationCountAnnotation] = existingRecreateCountStr
+	// add interval creation at the same point where we add the monitors
+	startClusterOperatorMonitoring(ctx, m, configClient)
+	return nil
+}
+
+func (m *Monitor) Stop(ctx context.Context, beginning, end time.Time) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.stopFn == nil {
+		return fmt.Errorf("monitor not started")
 	}
+	m.stopFn()
+	m.stopFn = nil
+
+	// we don't want this method to return until all te additional recorders and invariants have completed processing.
+	// to do this correctly, we need closure channels or some kind of mechanism.
+	// rather than properly wire this through, we'll wait until the backend disruption consumers timeout after the producer
+	// close.
+	// TODO once we have converted the backendsamplers to invariant tests, we can properly wait for completion
+	time.Sleep(70 * time.Second)
```

Comment:
This is the meat of what I added with an explanation as to why.  We can refine/fix during the refactor, but trying to wire it all properly now will be massive.